docs/migration.rst

```diff
@@ -0,0 +1,249 @@
+.. _migration:
+
+==============================
+Migrating to Shapely 1.8 / 2.0
+==============================
+
+Shapely 1.8.0 is a transitional version introducing several warnings in
+preparation of the upcoming changes in 2.0.0.
+
+Shapely 2.0.0 will be a major release with a refactor of the internals with
+considerable performance improvements (based on the developments in the
+`PyGEOS <https://github.com/pygeos/pygeos>`__ package), along with several
+breaking changes.
+
+This guide gives an overview of the most important changes with details
+on what will change in 2.0.0, how we warn for this in 1.8.0, and how
+you can update your code to be future-proof.
+
+For more background, see
+`RFC 1: Roadmap for Shapely 2.0 <https://github.com/shapely/shapely-rfc/pull/1>`__.
+
+.. contents:: Table of Contents
+
+
+Geometry objects will become immutable
+======================================
+
+Geometry objects will become immutable in version 2.0.0.
+
+In Shapely 1.x, some of the geometry classes are mutable, meaning that you
+can change their coordinates in-place. Illustrative code::
+
+    >>> from shapely.geometry import LineString
+    >>> line = LineString([(0,0), (2, 2)])
+    >>> print(line)
+    LINESTRING (0 0, 2 2)
+
+    >>> line.coords = [(0, 0), (10, 0), (10, 10)]
+    >>> print(line)
+    LINESTRING (0 0, 10 0, 10 10)
+
+In Shapely 1.8, this will start raising a warning::
+
+    >>> line.coords = [(0, 0), (10, 0), (10, 10)]
+    ShapelyDeprecationWarning: Setting the 'coords' to mutate a Geometry
+    in place is deprecated, and will not be possible any more in Shapely 2.0
+
+and starting with version 2.0.0, all geometry objects will become immutable
+(as a consequence, they will also become hashable and thefore usable as, for
+example, dictionary keys).
+
+**How do I update my code?** There is no direct alternative for mutating the
+coordinates of an existing geometry, except for creating a new geometry
+object with the new coordinates.
+
+
+Setting custom attributes
+-------------------------
+
+Another consequence of the geometry objects becoming immutable is that
+assigning custom attributes, which currently works, will no longer be possible.
+
+Currently you can do::
+
+    >>> line.name = "my_geometry"
+    >>> line.name
+    'my_geometry'
+
+This doesn't raise a warning in Shapely 1.8, but will start raising an
+AttributeError in Shapely 2.0.
```

Comment:
I realized this while writing this guide, but we could actually still add a warning for this. That might be useful?

Reply:
Looking into it, what I wrote here is actually not completely correct. So PyGEOS objects are indeed fully immutable (a C extension type):

```python
In [1]: import pygeos

In [2]: p = pygeos.Geometry("POINT(1 1)")

In [3]: p
Out[3]: <pygeos.Geometry POINT (1 1)>

In [4]: p.name = "test"
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
<ipython-input-4-3cf5ac6b0436> in <module>
----> 1 p.name = "test"

AttributeError: 'pygeos.lib.Geometry' object has no attribute 'name'
```

But when using the `shapely-2.0` branch, we actually create *python subclasses* for each geometry type (and register those to pygeos), and those subclasses are still mutable:

```python
In [1]: import pygeos

In [2]: p = pygeos.Geometry("POINT(1 1)")

In [3]: p
Out[3]: <shapely.geometry.Point POINT (1 1)>

In [4]: p.name = "test"

In [5]: p.name
Out[5]: 'test'
```

So in the end, it's not an inevitable consequence of the refactor, but we still have the _choice_ whether to allow or disallow this. 
I made a PR https://github.com/Toblerity/Shapely/pull/1181 to illustrate how we can disallow this in Shapely 2.0, but to discuss of course whether we want this or not.

Reply:
And also opened a PR to deprecate it for Shapely 1.8: https://github.com/Toblerity/Shapely/pull/1183